
### Завдання 1: Виклик LLM з базовим промптом

Створіть можливість викликати LLM зі звичайним текстовим промптом.

Промпт має дозвляти отримати інформацію простою мовою на певну тему. В цьому завданні ми хочемо дізнатись про тему "Квантові обчислення".

Відповідь моделі повинна містити визначення, ключові переваги та поточні дослідження в цій галузі.

Обмежте відповідь до 200 символів і пропишіть в промпті аби відповідь була короткою (це зекономить Вам час і гроші на згенеровані токени).

В якості LLM можна скористатись як моделлю з HugginFace (рекомендую Mistral), так і ChatGPT4 або ChatGPT3. В обох випадках треба імпортувати потрібну "обгортку" (тобто клас, який дозволить ініціювати модель) з LangChain для виклику LLM за API, а також зчитати особистий токен з файла, наприклад, `creds.json`, який розміщений у Вас локально і Ви НЕ здаєте його в ДЗ і НЕ комітите в git 😏

Встановіть своє значення температури на свій розсуд (тут немає правильного чи неправильного значення) і напишіть, чому ви обрали саме таке значення для цього завдання.  

Запити можна робити як українською, так і англійською - орієнтуйтесь на те, де і чи хочете ви потім лишити цей проєкт і відповідна яка мова буде пасувати більше. В розвʼязках промпти - українською.

In [134]:
import os

# Записав hugging face access key в окремий файл
file = open('credentials.txt').read()
HUGGING_FACE_ACCESS_KEY = file.split("\n")[0].split("=")[1]

os.environ["HF_TOKEN"] = HUGGING_FACE_ACCESS_KEY

In [2]:
!pip install -U langchain langchain-core langsmith langchain-huggingface

Defaulting to user installation because normal site-packages is not writeable


In [29]:
from huggingface_hub import InferenceClient
import langchain_huggingface
from langchain_huggingface import HuggingFaceEndpoint
from langchain_core.runnables import RunnableSequence

In [95]:
''' Виставляю температуру 0.1, для отримання найбільш релевантних відповідей, припустимо що я працюю дослідником в 
    науковому виданні та інформацію використовуватимуть студенти в навчальних цілях. Тому потрібно, щоб студенти
    "довіряли" даній інформації.
'''
overal_temperature = 0.1 

mistral_llm = HuggingFaceEndpoint(
    repo_id="mistralai/Mistral-7B-Instruct-v0.3",
    task="text-generation",
    temperature=overal_temperature,
    max_new_tokens=800
)

Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


In [27]:
response = mistral_llm.invoke('''Відповідай українською, мені потрібна лише відповідь, без коментарів. 
                    Дай визначення квантовим обчисленням. 
                    Наведи ключові переваги та поточні дослідження. 
                    Будь конкретним, опиши інформацію стисло.''')

/Users/bilukhan/Library/Python/3.12/lib/python/site-packages/huggingface_hub/utils/_deprecation.py:131: FutureWarning: 'post' (from 'huggingface_hub.inference._client') is deprecated and will be removed from version '0.31.0'. Making direct POST requests to the inference server is not supported anymore. Please use task methods instead (e.g. `InferenceClient.chat_completion`). If your use case is not supported, please open an issue in https://github.com/huggingface/huggingface_hub.
  warnings.warn(warning_message, FutureWarning)


In [28]:
print(response)



Квантове обчислення - це область комп'ютерних наук, яка використовує квантові ефекти для обчислень, щоб досягти швидкості, яка перевищує класичну обчислювальну швидкість.

Ключові переваги квантового обчислення:
1. Швидкість: Квантові комп'ютери можуть обчислювати за експоненціально меншу кількість кроків, ніж класичні комп'ютери.
2. енергоефективність: Квантові комп'ютери можуть використовувати менше енергії, ніж класичні комп'утери, для виконання одних і тих


### Завдання 2: Створення параметризованого промпта для генерації тексту
Тепер ми хочемо оновити попередній фукнціонал так, аби в промпт ми могли передавати тему як параметр. Для цього скористайтесь `PromptTemplate` з `langchain` і реалізуйте параметризований промпт та виклик моделі з ним.

Запустіть оновлений функціонал (промпт + модел) для пояснень про теми
- "Баєсівські методи в машинному навчанні"
- "Трансформери в машинному навчанні"
- "Explainable AI"

Виведіть результати відпрацювання моделі на екран.

In [43]:
from langchain_core.prompts import PromptTemplate

prompt = PromptTemplate(
    input_variables=["topic"],
    template="Відповідай українською, мені потрібна лише відповідь, без коментарів. Дай визначення {topic}. Наведи ключові переваги та поточні дослідження. Будь конкретним, опиши інформацію стисло."
)

In [47]:
prompt_input = prompt.format(topic="Баєсівські методи в машинному навчанні")

print(mistral_llm.invoke(prompt_input))



Баєсівські методи в машинному навчанні - це підхід до вирішення задач машинного навчання, який базується на теорії ймовірностей та припущенні про те, що всі параметри моделі є випадковими з певними а priori розподілами.

Ключові переваги Баєсівських методів:

1. **Природна інтерпретація**: Баєсівські методи дозволяють інтерпретувати результати як ймовірнісні розподіли, що робить їх привабливими для багатьох задач, де необхідно визначити ступінь довіри до результатів.
2. **


/Users/bilukhan/Library/Python/3.12/lib/python/site-packages/huggingface_hub/utils/_deprecation.py:131: FutureWarning: 'post' (from 'huggingface_hub.inference._client') is deprecated and will be removed from version '0.31.0'. Making direct POST requests to the inference server is not supported anymore. Please use task methods instead (e.g. `InferenceClient.chat_completion`). If your use case is not supported, please open an issue in https://github.com/huggingface/huggingface_hub.
  warnings.warn(warning_message, FutureWarning)


In [48]:
prompt_input = prompt.format(topic="Трансформери в машинному навчанні")

print(mistral_llm.invoke(prompt_input))



Трансформери (Transformer) - це архітектура моделі машинного навчання, яка використовується для задач машинного перекладу, аудіоперекладу та інших задач, що вимагають обробки послідовностей даних.

Основна перевага Трансформерів полягає в тому, що вони можуть ефективно обробляти довгі послідовності даних, не втрачаючи інформації, як це часто відбувається при використанні рекурентних нейронних мереж (RNN).

Ключовими дослідженнями в області Трансформерів є:

1. Розширення Трансформерів


/Users/bilukhan/Library/Python/3.12/lib/python/site-packages/huggingface_hub/utils/_deprecation.py:131: FutureWarning: 'post' (from 'huggingface_hub.inference._client') is deprecated and will be removed from version '0.31.0'. Making direct POST requests to the inference server is not supported anymore. Please use task methods instead (e.g. `InferenceClient.chat_completion`). If your use case is not supported, please open an issue in https://github.com/huggingface/huggingface_hub.
  warnings.warn(warning_message, FutureWarning)


In [69]:
prompt_input = prompt.format(topic="Explainable AI")

print(mistral_llm.invoke(prompt_input))

/Users/bilukhan/Library/Python/3.12/lib/python/site-packages/huggingface_hub/utils/_deprecation.py:131: FutureWarning: 'post' (from 'huggingface_hub.inference._client') is deprecated and will be removed from version '0.31.0'. Making direct POST requests to the inference server is not supported anymore. Please use task methods instead (e.g. `InferenceClient.chat_completion`). If your use case is not supported, please open an issue in https://github.com/huggingface/huggingface_hub.
  warnings.warn(warning_message, FutureWarning)




 Explainable AI (XAI) - це підхід до штучного інтелекту, який надає можливість людям розуміти та інтерпретувати рішення, які приймає штучний інтелекту. Це дозволяє зрозуміти, як алгоритми приймають рішення, що допомагає збільшити довіру до штучного інтелекту та зменшити ризики, пов'язані з непрозорими рішеннями.

 Ключові переваги XAI:

1. Збільшення довіри до штучного інтелекту: люди більш довіряють рішенням, які вони можуть розуміти.
2. Зменшення ри




### Завдання 3: Використання агента для автоматизації процесів
Створіть агента, який допоможе автоматично шукати інформацію про останні наукові публікації в різних галузях. Наприклад, агент має знайти 5 останніх публікацій на тему штучного інтелекту.

**Кроки:**
1. Налаштуйте агента типу ReAct в LangChain для виконання автоматичних запитів.
2. Створіть промпт, який спрямовує агента шукати інформацію в інтернеті або в базах даних наукових публікацій.
3. Агент повинен видати список публікацій, кожна з яких містить назву, авторів і короткий опис.

Для взаємодії з пошуком там необхідно створити `Tool`. В лекції ми використовували `serpapi`. Можна продовжити користуватись ним, або обрати інше АРІ для пошуку (вони в тому числі є безкоштовні). Перелік різних АРІ, доступних в langchain, і орієнтир по вартості запитів можна знайти в окремому документі [тут](https://hannapylieva.notion.site/API-12994835849480a69b2adf2b8441cbb3?pvs=4).

Лишаю також нижче приклад використання одного з безкоштовних пошукових АРІ - DuckDuckGo (не потребує створення токена!)  - можливо він вам сподобається :)


In [70]:
!pip install -q langchain_community duckduckgo_search

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [100]:
from langchain.utilities import DuckDuckGoSearchAPIWrapper

search = DuckDuckGoSearchAPIWrapper()

search.run("Obama's first name?")

"The White House, official residence of the president of the United States, in 2008. The president of the United States is the head of state and head of government of the United States, [1] indirectly elected to a four-year term via the Electoral College. [2] Under the U.S. Constitution, the officeholder leads the executive branch of the federal government and is the commander-in-chief of the ... Obama's father, Barack Obama, Sr., was a teenage goatherd in rural Kenya, won a scholarship to study in the United States, and eventually became a senior economist in the Kenyan government.Obama's mother, S. Ann Dunham, grew up in Kansas, Texas, and Washington state before her family settled in Honolulu.In 1960 she and Barack Sr. met in a Russian language class at the University of Hawaii ... The Irish Sun, It's a fake Barack Obama's brother posts forged document he claims is ex-president's 'Kenyan birth certificate,' March 11, 2017 Salon, Orly Taitz is at it again , Sept. 4, 2009 Jimmy, the f

In [104]:
!pip install -q langchain_ollama

In [105]:
from langchain import hub
from langchain.agents import load_tools
from langchain.agents import Tool, AgentExecutor, AgentType, create_react_agent, initialize_agent
from langchain_ollama.llms import OllamaLLM

На даному етапі я використаю іншу модель `Gemma2:27b`, оскільки використав безкоштовні кредити HuggingFace

In [141]:
model=file.split("\n")[1].split("=")[1]
url = file.split("\n")[2].split("=")[1]
gemma = OllamaLLM(model=model, base_url=url, temperature=overal_temperature)

In [143]:
search_tool = Tool(
    name="DuckDuckGo Search",
    func=search.run,
    description="A search engine. Useful for when you need to answer questions about current events. Input should be a search query."
)

prompt = hub.pull("hwchase17/react")

agent = create_react_agent(gemma, [search_tool], prompt)
agent_executor = AgentExecutor(agent=agent, tools=[search_tool], verbose=True, handle_parsing_errors=True)

/Users/bilukhan/Library/Python/3.12/lib/python/site-packages/langsmith/client.py:278: LangSmithMissingAPIKeyWarning: API key must be provided when using hosted LangSmith API
  warnings.warn(


In [153]:
response = agent_executor.invoke({"input": "Знайди 5 найновіших публікацій на тему штучний інтелект, які містять інформацію про автора. Виведи публікації списком, включаючи назву, автора та короткий опис публікації. Виведи результат українською, це дуже важливо."})



> Entering new AgentExecutor chain...
Thought: Щоб знайти найновіші публікації з штучного інтелекту з інформацією про авторів, мені потрібно скористатися пошуковою системою DuckDuckGo.

Action: DuckDuckGo Search
Action Input: "найновіші публікації штучний інтелект автор"CEO ІТ-компанії DevRain, автор книжки «ChatGPT, DALL·E, Midjourney. Як генеративний штучний інтелект змінює світ» Олександр Краковецький поділився в соцмережах добіркою книжок про штучний інтелект українською мовою. Автор романів зі штучним інтелектом — це складні програмні інструменти, які використовують штучний інтелект, зокрема обробку природної мови (NLP) і алгоритми машинного навчання, щоб допомогти авторам у створенні письмового вмісту. Дізнайтеся про найцікавіші книжкові новинки 2025 року у сферах бізнесу, штучного інтелекту, кібербезпеки та лідерства. Огляд must-read видань для професійного та особистого розвитку. Минулого року я публікував добірку топ-10 книг про штучний інтелект. Однак за останній рік вийшли

In [154]:
print(response['output'])

1. **Назва:** AI Day 2025
   **Автор:** Zaxid.net
   **Короткий опис:** Перша в Україні конференція, повністю присвячена застосуванню ШІ в бізнесі. Дискусії заходу були присвячені впливу штучного інтелекту на економіку та юридичні аспекти його використання.

2. **Назва:** В Україні навчають штучний інтелект ухвалювати рішення на полі бою
   **Автор:** Reuters
   **Короткий опис:** Стаття про українські розробки в галузі військового застосування штучного інтелекту.

3. **Назва:** Штучний інтелект 2025 рік
   **Автор:** Forbes Ukraine
   **Короткий опис:** Огляд книжкових новинок 2025 року, які відображають сучасні виклики у світі технологій, бізнесу та управління.

4. **Назва:** Генеральний директор Google DeepMind прогнозує появу AGI
   **Автор:** Галина Михайлова
   **Короткий опис:**  Прогноз про появу штучного генералізованого інтелекту (AGI), який зможе виконувати будь-які інтелектуальні завдання на рівні людини.

5. **Назва:** Штучний інтелект може бути дизайнером інтер'єрів
   **



### Завдання 4: Створення агента-помічника для вирішення бізнес-задач

Створіть агента, який допомагає вирішувати задачі бізнес-аналітики. Агент має допомогти користувачу створити прогноз по продажам на наступний рік враховуючи рівень інфляції і погодні умови. Агент має вміти використовувати Python і ходити в інтернет аби отримати актуальні дані.

**Кроки:**
1. Налаштуйте агента, який працюватиме з аналітичними даними, заданими текстом. Користувач пише

```
Ми експортуємо апельсини з Бразилії. В 2021 експортували 200т, в 2022 - 190т, в 2023 - 210т, в 2024 який ще не закінчився - 220т. Зроби оцінку скільки ми зможемо експортувати апельсинів в 2025 враховуючи погодні умови в Бразилії і попит на апельсини в світі виходячи з економічної ситуації.
```

2. Створіть запит до агента, що містить чітке завдання – видати результат бізнес аналізу або написати, що він не може цього зробити і запит користувача (просто може бути все одним повідомлленням).

3. Запустіть агента і проаналізуйте результати. Що можна покращити?


In [202]:
!pip install -q langchain_experimental numexpr

In [203]:
from langchain.chains import LLMMathChain
from langchain_experimental.utilities import PythonREPL

In [204]:
problem_chain = LLMMathChain.from_llm(llm=gemma)
math_tool = Tool.from_function(name="Calculator", 
                              func=problem_chain.run,
                              description="Useful for when you need to answer questions about math. This tool is only for math questions and nothing else. Only input math expressions.")

In [231]:
python_repl = PythonREPL()
python_tool = Tool(
    name="python_interpreter",
    func=python_repl.run,
    description="A Python shell. Use this to execute python commands. Input should be a valid python command. If you want to see the output of a value, you necessarily should print it out with `print(...)`. Otherwise you won't see the result! It's very important."
)

In [232]:
tools = [search_tool, math_tool, python_tool]

agent = create_react_agent(gemma, tools, prompt)
agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True, handle_parsing_errors=True)

In [236]:
response = agent_executor.invoke({"input": "Ми експортуємо апельсини з Бразилії. В 2021 експортували 200т, в 2022 - 190т, в 2023 - 210т, в 2024 який ще не закінчився - 220т. Зроби оцінку скільки ми зможемо експортувати апельсинів в 2025 враховуючи погодні умови в Бразилії і попит на апельсини в світі виходячи з економічної ситуації. Відповідай українсько.!"})



> Entering new AgentExecutor chain...
Thought: This question requires analyzing trends and making a prediction based on several factors. I need more information about weather patterns in Brazil and global orange demand. 

Action: DuckDuckGo Search
Action Input: "weather patterns in brazil 2025 forecast"Dry weather in southernmost Brazil and Argentina isn't likely to impact production expectations. But that could change, particularly in Argentina. ... 2025 forecast: Wet spring and tight planting windows. Apr 7, 2025 | 3 Min Read. Will a pivot to corn pay off? Apr 7, 2025 | 3 Min Read. Aug 26 - Aug 28, 2025. Dynamic Weather Pattern Across Brazil/Argentina Continues. Upper Air Pattern Changes: Wetter Argentina ... 02/12/2025, 12:06 pm EST. Show all. 02/10/2025, 5:19 am EST. Fig. 1-2: February 1-7, 2025, South America temperature anomalies and attendant 500 MB ... Global Soil Moisture Observation Trend/Forecast 03/06/2025, 1:45 pm EST; Update on the ... Hotter and drier weather shifts ac

In [237]:
print(response['output'])

Враховуючи зростання попиту на апельсини та прогнозовані складні погодні умови в Бразилії, ми можемо оцінити експорт апельсинів в 2025 році на рівні близько 231 тисяч тонн.


Отже, модель врахувала погодні умови, так як зробила запит на отримання інформації про прогноз погоди в Бразилії у 2025 році. Також врахувала попит на апельсини, тому також здійснила запит. 

* <i>Thought: The weather patterns in Brazil for 2025 are predicted to be challenging with droughts and intense rainfall. This could negatively impact orange production. I need information about the global demand for oranges.</i>
* <i>Thought: The global demand for oranges is projected to grow, particularly in Europe. However, the challenging weather conditions in Brazil could significantly impact production.</i>

Модель побачила що попит на апельсини зростатиме, зокрема у Європі, проте погодні умови будуть дещо нестабільними. Тому вона припустила що приріст буде 10% та передбачила експорт у 231 тисячу тон.

Що можна покращити: передати усю базу експорту апельсині за днями, передати в агента щоденний моніторинг, а саме OpenWeatherMap tool та написати кастомний предиктор, який передбачатиме експорт щодня.